### PR TITLE
Use `tokio::sync::Mutex` to make lock get across yield point

### PR DESCRIPTION
**Pull request by weihanglo**
## Motivation and Context

Address #1674, to remove the temporarily allowed lint rule [`await_holding_lock`](https://rust-lang.github.io/rust-clippy/master/index.html#await_holding_lock).

## Description

[`parking_lot::Mutex`](https://docs.rs/parking_lot/0.12.1/parking_lot/type.Mutex.html) cannot be used across yield point, so replace with [`tokio::sync::Mutex`](https://docs.rs/tokio/1.20.1/tokio/sync/struct.Mutex.html).

Note that `tokio::sync::Mutex` is more expensive than sync Mutex, so if we can figure out a better way to fix this, I am happy to.

I also remove the `size_hint` impl since it's not easy to use an async lock inside a sync function. Any better approach to put it back is welcome.

## Testing

Ran some apachebench and it seems like the new one is rougly the same as the old one. I would say we can settle on this and see if we can improve it in the future.

<details>

<summary>Benchamrk results after this patch</summary>

### After (debug mode)

<table >
<tr ><th colspan=2 bgcolor=white>Server Software:</th><td colspan=2 bgcolor=white></td></tr>
<tr ><th colspan=2 bgcolor=white>Server Hostname:</th><td colspan=2 bgcolor=white>127.0.0.1</td></tr>
<tr ><th colspan=2 bgcolor=white>Server Port:</th><td colspan=2 bgcolor=white>13734</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Path:</th><td colspan=2 bgcolor=white>/radio</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Length:</th><td colspan=2 bgcolor=white>Variable</td></tr>
<tr ><th colspan=2 bgcolor=white>Concurrency Level:</th><td colspan=2 bgcolor=white>100</td></tr>
<tr ><th colspan=2 bgcolor=white>Time taken for tests:</th><td colspan=2 bgcolor=white>56.422 seconds</td></tr>
<tr ><th colspan=2 bgcolor=white>Complete requests:</th><td colspan=2 bgcolor=white>1000</td></tr>
<tr ><th colspan=2 bgcolor=white>Failed requests:</th><td colspan=2 bgcolor=white>0</td></tr>
<tr ><th colspan=2 bgcolor=white>Total transferred:</th><td colspan=2 bgcolor=white>3023157030 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>HTML transferred:</th><td colspan=2 bgcolor=white>3023061030 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>Requests per second:</th><td colspan=2 bgcolor=white>17.72</td></tr>
<tr ><th colspan=2 bgcolor=white>Transfer rate:</th><td colspan=2 bgcolor=white>52325.57 kb/s received</td></tr>
<tr ><th bgcolor=white colspan=4>Connection Times (ms)</th></tr>
<tr ><th bgcolor=white>&nbsp;</th> <th bgcolor=white>min</th>   <th bgcolor=white>avg</th>   <th bgcolor=white>max</th></tr>
<tr ><th bgcolor=white>Connect:</th><td bgcolor=white>    0</td><td bgcolor=white>    0</td><td bgcolor=white>    7</td></tr>
<tr ><th bgcolor=white>Processing:</th><td bgcolor=white> 1610</td><td bgcolor=white> 4625</td><td bgcolor=white>20552</td></tr>
<tr ><th bgcolor=white>Total:</th><td bgcolor=white> 1610</td><td bgcolor=white> 4625</td><td bgcolor=white>20559</td></tr>
</table>

### After (release mode)

<table >
<tr ><th colspan=2 bgcolor=white>Server Software:</th><td colspan=2 bgcolor=white></td></tr>
<tr ><th colspan=2 bgcolor=white>Server Hostname:</th><td colspan=2 bgcolor=white>127.0.0.1</td></tr>
<tr ><th colspan=2 bgcolor=white>Server Port:</th><td colspan=2 bgcolor=white>13734</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Path:</th><td colspan=2 bgcolor=white>/radio</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Length:</th><td colspan=2 bgcolor=white>Variable</td></tr>
<tr ><th colspan=2 bgcolor=white>Concurrency Level:</th><td colspan=2 bgcolor=white>100</td></tr>
<tr ><th colspan=2 bgcolor=white>Time taken for tests:</th><td colspan=2 bgcolor=white>62.987 seconds</td></tr>
<tr ><th colspan=2 bgcolor=white>Complete requests:</th><td colspan=2 bgcolor=white>1000</td></tr>
<tr ><th colspan=2 bgcolor=white>Failed requests:</th><td colspan=2 bgcolor=white>0</td></tr>
<tr ><th colspan=2 bgcolor=white>Total transferred:</th><td colspan=2 bgcolor=white>3006106880 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>HTML transferred:</th><td colspan=2 bgcolor=white>3006010880 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>Requests per second:</th><td colspan=2 bgcolor=white>15.88</td></tr>
<tr ><th colspan=2 bgcolor=white>Transfer rate:</th><td colspan=2 bgcolor=white>46607.21 kb/s received</td></tr>
<tr ><th bgcolor=white colspan=4>Connection Times (ms)</th></tr>
<tr ><th bgcolor=white>&nbsp;</th> <th bgcolor=white>min</th>   <th bgcolor=white>avg</th>   <th bgcolor=white>max</th></tr>
<tr ><th bgcolor=white>Connect:</th><td bgcolor=white>    0</td><td bgcolor=white>    0</td><td bgcolor=white>    8</td></tr>
<tr ><th bgcolor=white>Processing:</th><td bgcolor=white> 2518</td><td bgcolor=white> 4930</td><td bgcolor=white>20682</td></tr>
<tr ><th bgcolor=white>Total:</th><td bgcolor=white> 2518</td><td bgcolor=white> 4930</td><td bgcolor=white>20690</td></tr>
</table>

</details>

<details>

<summary>Benchamrk results before this patch</summary>

### Before (debug mode)

<table >
<tr ><th colspan=2 bgcolor=white>Server Software:</th><td colspan=2 bgcolor=white></td></tr>
<tr ><th colspan=2 bgcolor=white>Server Hostname:</th><td colspan=2 bgcolor=white>127.0.0.1</td></tr>
<tr ><th colspan=2 bgcolor=white>Server Port:</th><td colspan=2 bgcolor=white>13734</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Path:</th><td colspan=2 bgcolor=white>/radio</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Length:</th><td colspan=2 bgcolor=white>Variable</td></tr>
<tr ><th colspan=2 bgcolor=white>Concurrency Level:</th><td colspan=2 bgcolor=white>100</td></tr>
<tr ><th colspan=2 bgcolor=white>Time taken for tests:</th><td colspan=2 bgcolor=white>61.400 seconds</td></tr>
<tr ><th colspan=2 bgcolor=white>Complete requests:</th><td colspan=2 bgcolor=white>1000</td></tr>
<tr ><th colspan=2 bgcolor=white>Failed requests:</th><td colspan=2 bgcolor=white>0</td></tr>
<tr ><th colspan=2 bgcolor=white>Total transferred:</th><td colspan=2 bgcolor=white>2951546400 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>HTML transferred:</th><td colspan=2 bgcolor=white>2951450400 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>Requests per second:</th><td colspan=2 bgcolor=white>16.29</td></tr>
<tr ><th colspan=2 bgcolor=white>Transfer rate:</th><td colspan=2 bgcolor=white>46944.30 kb/s received</td></tr>
<tr ><th bgcolor=white colspan=4>Connection Times (ms)</th></tr>
<tr ><th bgcolor=white>&nbsp;</th> <th bgcolor=white>min</th>   <th bgcolor=white>avg</th>   <th bgcolor=white>max</th></tr>
<tr ><th bgcolor=white>Connect:</th><td bgcolor=white>    0</td><td bgcolor=white>    1</td><td bgcolor=white>    9</td></tr>
<tr ><th bgcolor=white>Processing:</th><td bgcolor=white> 2527</td><td bgcolor=white> 4875</td><td bgcolor=white>21891</td></tr>
<tr ><th bgcolor=white>Total:</th><td bgcolor=white> 2527</td><td bgcolor=white> 4876</td><td bgcolor=white>21900</td></tr>
</table>

#### Before (release mode)

<table >
<tr ><th colspan=2 bgcolor=white>Server Software:</th><td colspan=2 bgcolor=white></td></tr>
<tr ><th colspan=2 bgcolor=white>Server Hostname:</th><td colspan=2 bgcolor=white>127.0.0.1</td></tr>
<tr ><th colspan=2 bgcolor=white>Server Port:</th><td colspan=2 bgcolor=white>13734</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Path:</th><td colspan=2 bgcolor=white>/radio</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Length:</th><td colspan=2 bgcolor=white>Variable</td></tr>
<tr ><th colspan=2 bgcolor=white>Concurrency Level:</th><td colspan=2 bgcolor=white>100</td></tr>
<tr ><th colspan=2 bgcolor=white>Time taken for tests:</th><td colspan=2 bgcolor=white>63.088 seconds</td></tr>
<tr ><th colspan=2 bgcolor=white>Complete requests:</th><td colspan=2 bgcolor=white>1000</td></tr>
<tr ><th colspan=2 bgcolor=white>Failed requests:</th><td colspan=2 bgcolor=white>0</td></tr>
<tr ><th colspan=2 bgcolor=white>Total transferred:</th><td colspan=2 bgcolor=white>2961776490 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>HTML transferred:</th><td colspan=2 bgcolor=white>2961680490 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>Requests per second:</th><td colspan=2 bgcolor=white>15.85</td></tr>
<tr ><th colspan=2 bgcolor=white>Transfer rate:</th><td colspan=2 bgcolor=white>45846.50 kb/s received</td></tr>
<tr ><th bgcolor=white colspan=4>Connection Times (ms)</th></tr>
<tr ><th bgcolor=white>&nbsp;</th> <th bgcolor=white>min</th>   <th bgcolor=white>avg</th>   <th bgcolor=white>max</th></tr>
<tr ><th bgcolor=white>Connect:</th><td bgcolor=white>    0</td><td bgcolor=white>    0</td><td bgcolor=white>    8</td></tr>
<tr ><th bgcolor=white>Processing:</th><td bgcolor=white> 2682</td><td bgcolor=white> 4809</td><td bgcolor=white>21418</td></tr>
<tr ><th bgcolor=white>Total:</th><td bgcolor=white> 2682</td><td bgcolor=white> 4809</td><td bgcolor=white>21426</td></tr>
</table>

</details>

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by weihanglo** [comment]
cc @crisidev

File: rust-runtime/aws-smithy-http-server-python/src/types.rs [hunk 1]
```diff
@@ -6,19 +6,20 @@
...
 
 use std::{
+    future::Future,
     pin::Pin,
     sync::Arc,
     task::{Context, Poll},
 };
 
 use bytes::Bytes;
-use parking_lot::Mutex;
```
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server-python/src/types.rs:15]: I think you can get rid of the `parking_lot` dependency entirely, no?

File: rust-runtime/aws-smithy-http-server-python/src/types.rs [hunk 2]
```diff
@@ -288,22 +289,20 @@ impl futures::stream::Stream for ByteStream {
...
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        let mut stream = self.0.lock();
-        Pin::new(&mut *stream).poll_next(cx)
-    }
-
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        self.0.lock().size_hint()
+        let stream = self.0.lock();
+        tokio::pin!(stream);
```
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server-python/src/types.rs:293]: I need to really dig into tokio. I did not know about `tokio::pin!`. Very nice @weihanglo !!

**Review by crisidev — APPROVED** [review]
👏🏼

File: rust-runtime/aws-smithy-http-server-python/src/types.rs (hunk 2 above)
Comment (weihanglo) [reply rust-runtime/aws-smithy-http-server-python/src/types.rs:293]: I myself need to really dig into `std::pin::Pin` as well 😆

File: rust-runtime/aws-smithy-http-server-python/src/types.rs (hunk 1 above)
Comment (weihanglo) [reply rust-runtime/aws-smithy-http-server-python/src/types.rs:15]: I guess not 😞 

https://github.com/awslabs/smithy-rs/blob/30d985ad9d39849bb40705d244413eae2764ba7f/rust-runtime/aws-smithy-http-server-python/src/server.rs#L10

**Review by crisidev — APPROVED** [review]

**Pull request merged**